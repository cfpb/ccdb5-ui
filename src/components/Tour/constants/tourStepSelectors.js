export const DOCUMENT_STEP_SELECTORS = {
  STEP_1: '.content__hero',
  STEP_2: '.m-pagination__btn-next',
  STEP_3: '.saved__search-panel div:nth-child(2)',
  STEP_4: '.back-to-results',
};

export const GEO_STEP_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '.map-shading', label: 'Map Shading' },
  STEP_13: { selector: '#tile-chart-map', label: 'Tile Map' },
  STEP_14: { selector: '.row-chart-section', label: 'Row Charts' },
  STEP_15: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};

export const LIST_COMPLAINTS_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '[data-tour="select-size"]', label: 'Show Number' },
  STEP_13: { selector: '[data-tour="select-sort"]', label: 'Sort Results' },
  STEP_14: {
    selector: '.refine-bar section.m-btn-group',
    label: 'Show Narratives',
  }, // read complaints
  STEP_15: { selector: '.cards-panel', label: 'Results' },
  STEP_16: { selector: '.to-detail', label: 'View Detailed Complaints' }, // complaint id
  STEP_17: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};

export const TRENDS_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '[data-tour="select-lens"]', label: 'Aggregate By' },
  STEP_13: {
    selector: '[data-tour="select-interval"]',
    label: 'Date Interval',
  },
  STEP_14: { selector: '.row-chart-section', label: 'Row Charts' },
  STEP_15: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};
