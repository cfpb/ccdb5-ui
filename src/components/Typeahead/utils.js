/**
 * Sets the options for the async typeahead
 *
 * @param {string} value - value typed into the typeahead
 * @param {Function} setOptions - function to set local array of options
 * @param {string} uri - the endpoint being hit by the fetch
 * @returns {void | Promise<object>} - Either nothing or a response with option object
 */
export function handleFetchSearch(value, setOptions, uri) {
  const num = value.toLowerCase();
  if (num === '') {
    setOptions([]);
    return;
  }
  return fetch(uri)
    .then((result) => result.json())
    .then((items) => {
      const options = items.map((item) => ({
        key: item,
        label: item,
        position: item.toLowerCase().indexOf(num),
        value,
      }));
      setOptions(options);
    });
}
