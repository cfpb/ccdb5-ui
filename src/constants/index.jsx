// Internal triggers
export const API_CALLED = 'API_CALLED'
export const MODAL_HID = 'MODAL_HID'
export const MODAL_SHOWN = 'MODAL_SHOWN'

// Modal Dialogs
export const MODAL_TYPE_DATA_EXPORT = 'MODAL_TYPE_DATA_EXPORT'
export const MODAL_TYPE_MORE_ABOUT = 'MODAL_TYPE_MORE_ABOUT'

// Useful constants
export const NARRATIVE_SEARCH_FIELD = 'complaint_what_happened'

// query manager flags
// These constants control how the query manager works
export const REQUERY_ALWAYS = 'REQUERY_ALWAYS'
export const REQUERY_HITS_ONLY = 'REQUERY_HITS_ONLY'
// default if not specified
export const REQUERY_NEVER = 'REQUERY_NEVER'

export const knownFilters = [
  'company',
  'company_public_response',
  'company_response',
  'consumer_consent_provided',
  'consumer_disputed',
  'issue',
  'product',
  'state',
  'submitted_via',
  'tag',
  'timely',
  'zip_code'
]

export const dateFilters = [
  'company_received_max',
  'company_received_min',
  'date_received_max',
  'date_received_min'
]

export const flagFilters = [
  'has_narrative'
]

export const SLUG_SEPARATOR = 'â€¢'

// Keyboard
export const VK_DOWN = 40
export const VK_ENTER = 14
export const VK_ESCAPE = 27
export const VK_RETURN = 13
export const VK_TAB = 9
export const VK_UP = 38

/// ////////////////////////////////////////////////////////////////////////////
// JSON Objects

export const THESE_UNITED_STATES = {
  AA: 'Armed Forces Americas',
  AE: 'Armed Forces Europe',
  AL: 'Alabama',
  AK: 'Alaska',
  AS: 'American Samoa',
  AP: 'Armed Forces Pacific',
  AZ: 'Arizona',
  AR: 'Arkansas',
  CA: 'California',
  CO: 'Colorado',
  CT: 'Connecticut',
  DE: 'Delaware',
  DC: 'District Of Columbia',
  FM: 'Federated States Of Micronesia',
  FL: 'Florida',
  GA: 'Georgia',
  GU: 'Guam',
  HI: 'Hawaii',
  ID: 'Idaho',
  IL: 'Illinois',
  IN: 'Indiana',
  IA: 'Iowa',
  KS: 'Kansas',
  KY: 'Kentucky',
  LA: 'Louisiana',
  ME: 'Maine',
  MH: 'Marshall Islands',
  MD: 'Maryland',
  MA: 'Massachusetts',
  MI: 'Michigan',
  MN: 'Minnesota',
  MS: 'Mississippi',
  MO: 'Missouri',
  MT: 'Montana',
  NE: 'Nebraska',
  NV: 'Nevada',
  NH: 'New Hampshire',
  NJ: 'New Jersey',
  NM: 'New Mexico',
  NY: 'New York',
  NC: 'North Carolina',
  ND: 'North Dakota',
  MP: 'Northern Mariana Islands',
  OH: 'Ohio',
  OK: 'Oklahoma',
  OR: 'Oregon',
  PW: 'Palau',
  PA: 'Pennsylvania',
  PR: 'Puerto Rico',
  RI: 'Rhode Island',
  SC: 'South Carolina',
  SD: 'South Dakota',
  TN: 'Tennessee',
  TX: 'Texas',
  UT: 'Utah',
  VT: 'Vermont',
  VI: 'Virgin Islands',
  VA: 'Virginia',
  WA: 'Washington',
  WV: 'West Virginia',
  WI: 'Wisconsin',
  WY: 'Wyoming'
}

export const TILE_MAP_STATES = [
  'AK',
  'AL',
  'AR',
  'AZ',
  'CA',
  'CO',
  'CT',
  'DC',
  'DE',
  'FL',
  'GA',
  'HI',
  'IA',
  'ID',
  'IL',
  'IN',
  'KS',
  'KY',
  'LA',
  'MA',
  'MD',
  'ME',
  'MI',
  'MN',
  'MO',
  'MS',
  'MT',
  'NC',
  'ND',
  'NE',
  'NH',
  'NJ',
  'NM',
  'NV',
  'NY',
  'OH',
  'OK',
  'OR',
  'PA',
  'RI',
  'SC',
  'SD',
  'TN',
  'TX',
  'UT',
  'VA',
  'VT',
  'WA',
  'WI',
  'WV',
  'WY'
]
