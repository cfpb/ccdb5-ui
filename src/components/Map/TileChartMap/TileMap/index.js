/* eslint complexity: ["error", 5] */
import * as d3 from 'd3';
import accessibility from 'highcharts/modules/accessibility';
import Highcharts from 'highcharts/highmaps';
import { STATE_TILES } from './constants';

const TEN_K = 10000;
const HUN_K = 100000;
const MILLION = 1000000;

const WHITE = '#ffffff';

/* ----------------------------------------------------------------------------
   Utility Functions */

/**
 * Creates N evenly spaced ranges in the data
 *
 * @param {Array} data - all of the states w/ displayValue, complaintCount, raw
 * @param {Array} colors - an array of colors
 * @returns {Array} floating point numbers that mark the max of each range
 */
export function makeScale(data, colors) {
  const allValues = data.map((datum) => datum.displayValue);
  const uniques = new Set(allValues);

  let scale = d3.scaleQuantile().range([WHITE, ...colors]);
  // This catches the condition where all the complaints are in one state
  if (uniques.size < colors.length) {
    scale = scale.domain([...uniques]);
  } else {
    scale = scale.domain(allValues);
  }

  return scale;
}

/**
 * Creates a shorter version of a number. 1,234 => 1.2K
 *
 * @param {number} value - the raw value
 * @returns {string} A string representing a shortened value
 */
export function makeShortName(value) {
  if (value < 1000) {
    return value.toLocaleString();
  } else if (value < TEN_K) {
    return (Math.floor(value / 100) / 10).toFixed(1) + 'K';
  } else if (value < MILLION) {
    return Math.floor(value / 1000) + 'K';
  }

  return (Math.floor(value / HUN_K) / 10).toFixed(1) + 'M';
}

/* ----------------------------------------------------------------------------
   Bin Functions */

/**
 * helper function to get the bins for legend and colors, etc.
 *
 * @param {Array} quantiles - floats that mark the max of each range
 * @param {Function} scale - scaling function for color
 * @returns {Array} the bins with bounds, name, and color
 */
export function getBins(quantiles, scale) {
  const rounds = quantiles.map((quant) => Math.round(quant));
  const ceils = quantiles.map((quant) => Math.ceil(quant));
  const mins = Array.from(new Set(rounds)).filter((round) => round > 0);

  const bins = [{ from: 0, color: WHITE, name: '≥ 0', shortName: '≥ 0' }];

  mins.forEach((minValue) => {
    // The color is the equivalent ceiling from the floor
    const idx = rounds.indexOf(minValue);

    const prefix = ceils[idx] === minValue ? '≥' : '>';
    const displayValue = minValue.toLocaleString();
    const shortened = makeShortName(minValue);

    bins.push({
      from: minValue,
      color: scale(ceils[idx]),
      name: `${prefix} ${displayValue}`,
      shortName: `${prefix} ${shortened}`,
    });
  });

  return bins;
}

/**
 * helper function to get the Per 1000 population bins for legend and colors
 *
 * @param {Array} quantiles - floats that mark the max of each range
 * @param {Function} scale - scaling function for color
 * @returns {Array} the bins with bounds, name, and color
 */
export function getPerCapitaBins(quantiles, scale) {
  const trunc100 = (num) => Math.floor(num * 100) / 100;

  const values = quantiles.map((val) => trunc100(val));
  const mins = Array.from(new Set(values)).filter((val) => val > 0);

  const bins = [{ from: 0, color: WHITE, name: '≥ 0', shortName: '≥ 0' }];

  mins.forEach((minValue) => {
    // The color is the equivalent quantile
    const idx = values.indexOf(minValue);

    const prefix = values[idx] === quantiles[idx] ? '≥' : '>';
    const displayValue = minValue.toFixed(2);
    const name = `${prefix} ${displayValue}`;
    bins.push({
      from: minValue,
      color: scale(quantiles[idx]),
      name,
      shortName: name,
    });
  });

  return bins;
}

/* ----------------------------------------------------------------------------
   Utility Functions 2 */
/**
 * @param {object} data - Data to process. add in state paths to the data obj
 * @param {Function} scale - scaling function for color
 * @returns {object} The processed data.
 */
export function processMapData(data, scale) {
  // Filter out any empty values just in case
  data = data.filter(function (row) {
    return Boolean(row.name);
  });

  const isFiltered = data.filter((obj) => obj.className === 'selected').length;
  data = data.map(function (obj) {
    const path = STATE_TILES[obj.name];
    let color = getColorByValue(obj.displayValue, scale);

    if (isFiltered && obj.className === 'deselected') {
      // update rgba opacity for selected state
      color = color.replace('1)', '0.5)');
    }

    if (obj.className !== 'selected' && color === WHITE) {
      // handle cases where value is empty or no color, so we can set the border
      obj.className = 'empty';
    }

    return {
      ...obj,
      color,
      path,
    };
  });

  return data;
}

/**
 * helper function to set the color.
 *
 * Highcharts could normally handle it, but it gets confused by values
 * less than 1 that are frequently encountered in perCapita
 *
 * Also, walk through the array backwards to pick up the most saturated
 * color. This helps the "only three values" case
 *
 * @param {number} value - the number of complaints or perCapita
 * @param {Function} scale - scaling function for color
 * @returns {string} color hex or rgb code for a color
 */
export function getColorByValue(value, scale) {
  if (!value) return WHITE;

  return scale(value);
}

/* ----------------------------------------------------------------------------
   Highcharts callbacks */

/**
 * callback function for reporting the series point in a voiceover text
 *
 * @param {object} point - the point in the series
 * @returns {string} the text to speak
 */
export function descriptionFormatter(point) {
  return `${point.fullName} ${point.displayValue}`;
}

/**
 * callback function for mouseout a point to remove hover class from tile label
 */
export function mouseoutPoint() {
  const name = '.tile-' + this.name;
  d3.select(name).classed('hover', false);
}

/**
 * callback function for mouseover point to add hover class to tile label
 */
export function mouseoverPoint() {
  const name = '.tile-' + this.name;
  d3.select(name).classed('hover', true);
}

/**
 * callback function to format the individual tiles in HTML
 *
 * @returns {string} html output
 */
export function tileFormatter() {
  const value = this.point.displayValue.toLocaleString();
  return (
    '<div class="highcharts-data-label-state tile-' +
    this.point.name +
    ' ' +
    this.point.className +
    ' ">' +
    '<span class="abbr">' +
    this.point.name +
    '</span>' +
    '<span class="value">' +
    value +
    '</span>' +
    '</div>'
  );
}

/**
 * callback function to format the tooltip in HTML
 *
 * @returns {string} html output
 */
export function tooltipFormatter() {
  const product = this.product
    ? '<div class="row u-clearfix">' +
      '<p class="u-float-left">Product with highest complaint volume</p>' +
      '<p class="u-right">' +
      this.product +
      '</p>' +
      '</div>'
    : '';

  const issue = this.issue
    ? '<div class="row u-clearfix">' +
      '<p class="u-float-left">Issue with highest complaint volume</p>' +
      '<p class="u-right">' +
      this.issue +
      '</p>' +
      '</div>'
    : '';

  const value = this.value.toLocaleString();
  const perCapita = this.perCapita
    ? '<div class="row u-clearfix">' +
      '<p class="u-float-left">Per 1000 population</p>' +
      '<p class="u-right">' +
      this.perCapita +
      '</p>' +
      '</div>'
    : '';

  return (
    '<div class="title">' +
    this.fullName +
    '</div>' +
    '<div class="row u-clearfix">' +
    '<p class="u-float-left">Complaints</p>' +
    '<p class="u-right">' +
    value +
    '</p>' +
    '</div>' +
    perCapita +
    product +
    issue
  );
}

/**
 * Draw a legend on a chart.
 *
 * @param {object} chart - A highchart chart.
 */
export function _drawLegend(chart) {
  const bins = chart.options.bins;
  let boxWidth = 65;
  const boxHeight = 17;
  let boxPadding = 5;

  const beCompact = chart.chartWidth < 600;
  if (beCompact) {
    boxWidth = 45;
    boxPadding = 1;
  }

  /* https://api.highcharts.com/class-reference/Highcharts.SVGRenderer#label
     boxes and labels for legend buckets */
  // main container
  const legendContainer = chart.renderer.g('legend-container').add();

  const legendText = chart.renderer
    .g('legend-title')
    .translate(boxPadding, 0)
    .add(legendContainer);
  // key
  chart.renderer
    .label('Key', 0, 0, null, null, null, true, false, 'legend-key')
    .add(legendText);

  // horizontal separator line
  const sepWidth = bins.length * (boxWidth + boxPadding);
  chart.renderer
    .path(['M', 0, 0, 'L', sepWidth, 0])
    .attr({
      class: 'separator',
      'stroke-width': 1,
      stroke: 'gray',
    })
    .translate(0, 25)
    .add(legendText);

  // what legend represents
  const labelTx =
    'Map shading: <span class="type">' +
    chart.options.legend.legendTitle +
    '</span>';
  chart.renderer
    .label(labelTx, 0, 28, null, null, null, true, false, 'legend-description')
    .add(legendText);

  // bars
  const legend = chart.renderer
    .g('legend__tile-map')
    .translate(7, 50)
    .add(legendContainer);

  for (let idx = 0; idx < bins.length; idx++) {
    const rend = chart.renderer
      .g(`g${idx}`)
      .translate(idx * (boxWidth + boxPadding), 0)
      .add(legend);

    const bin = bins[idx];

    chart.renderer
      .rect(0, 0, boxWidth, boxHeight)
      .attr({ fill: bin.color })
      .addClass('legend-box')
      .add(rend);

    chart.renderer
      .text(beCompact ? bin.shortName : bin.name, 0, boxHeight)
      .addClass('legend-text')
      .translate(3, -3)
      .add(rend);
  }
}

/* ----------------------------------------------------------------------------
   Tile Map class */

accessibility(Highcharts);

Highcharts.setOptions({
  lang: {
    thousandsSep: ',',
  },
});

const colors = [
  'rgba(212, 231, 230, 1)',
  'rgba(180, 210, 209, 1)',
  'rgba(158, 196, 195, 1)',
  'rgba(137, 182, 181, 1)',
  'rgba(112, 166, 165, 1)',
  'rgba(87, 150, 149, 1)',
];

/* ----------------------------------------------------------------------------
   Tile Map class */

class TileMap {
  constructor({ el, data, isPerCapita, events, height, hasTip, width }) {
    const scale = makeScale(data, colors);
    const quantiles = scale.quantiles();

    let bins, legendTitle;
    if (isPerCapita) {
      bins = getPerCapitaBins(quantiles, scale);
      legendTitle = 'Complaints per 1,000';
    } else {
      bins = getBins(quantiles, scale);
      legendTitle = 'Complaints';
    }

    data = processMapData(data, scale);

    const options = {
      accessibility: {
        description: '',
        screenReaderSection: {
          afterChartFormat: '',
          beforeChartFormat: '',
        },
      },

      bins,
      chart: {
        styledMode: true,
        height,
        width,
      },
      colorAxis: {
        dataClasses: bins,
        dataClassColor: 'category',
      },
      title: false,
      credits: false,
      legend: {
        enabled: false,
        legendTitle,
      },
      tooltip: {
        className: 'tooltip',
        enabled: hasTip,
        headerFormat: '',
        pointFormatter: tooltipFormatter,
        useHTML: true,
      },
      plotOptions: {
        series: {
          dataLabels: {
            enabled: true,
            formatter: tileFormatter,
            useHTML: true,
          },
          states: {
            hover: {
              enabled: hasTip,
            },
          },
        },
      },

      series: [
        {
          type: 'map',
          clip: false,
          data: data,
          accessibility: {
            description: legendTitle + ' in the United States',
            exposeAsGroupOnly: false,
            keyboardNavigation: { enabled: true },
            descriptionFormatter: descriptionFormatter,
          },
        },
      ],
    };

    // our custom passing of information
    if (events && hasTip) {
      options.plotOptions.series.events = events;
      options.plotOptions.series.point = {
        events: {
          mouseOver: mouseoverPoint,
          mouseOut: mouseoutPoint,
        },
      };
    }

    // to adjust for legend height
    const mapBreakpoints = [
      { width: 700, legendHeight: 20 },
      { width: 580, legendHeight: 25 },
      { width: 500, legendHeight: 35 },
      { width: 400, legendHeight: 60 },
      { width: 370, legendHeight: 70 },
    ];

    let legendHeight = 10;

    mapBreakpoints.forEach((item) => {
      if (width < item.width) {
        legendHeight = item.legendHeight;
      }
    });

    options.chart.marginRight = 0;
    options.chart.marginLeft = 0;
    options.chart.marginTop = legendHeight;
    options.chart.height += legendHeight;

    this.draw(el, options);
  }

  draw(el, options) {
    Highcharts.mapChart(el, options, _drawLegend);
  }
}

export default TileMap;
