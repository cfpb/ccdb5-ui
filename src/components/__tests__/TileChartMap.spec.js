import configureMockStore from 'redux-mock-store';
import ReduxTileChartMap, {
  mapDispatchToProps,
  mapStateToProps,
  TileChartMap,
} from '../Charts/TileChartMap';
import { Provider } from 'react-redux';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import TileMap from '../Charts/TileMap';
import * as utils from '../../utils';
import * as types from '../../constants';

jest.mock('../Charts/TileMap');

function setupSnapshot(printMode) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    map: {
      results: {
        state: [],
      },
    },
    query: {
      state: [],
    },
    view: {
      printMode,
    },
  });

  return renderer.create(
    <Provider store={store}>
      <ReduxTileChartMap />
    </Provider>
  );
}

describe('component: TileChartMap', () => {
  let mapDiv, redrawSpy, target;
  describe('initial state', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('renders without crashing', () => {
      const target = setupSnapshot(false);
      let tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders print mode without crashing', () => {
      const target = setupSnapshot(true);
      let tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });

    describe('when clicking a map node', () => {
      it('adds a map filter when it is not currently a filter', () => {
        target = shallow(<TileChartMap />);
        const mapEvent = { point: { abbr: 'FO', fullName: 'Foo Bar' } };
        const instance = target.instance();
        instance.stateFilters = ['CA'];
        instance.addState = jest.fn();
        instance._toggleState(mapEvent);
        expect(instance.addState).toHaveBeenCalledTimes(1);
      });

      it('removes a map filter when it is currently a filter', () => {
        target = shallow(<TileChartMap />);
        const mapEvent = { point: { abbr: 'FO', fullName: 'Foo Bar' } };
        const instance = target.instance();
        instance.stateFilters = ['FO'];
        instance.removeState = jest.fn();
        instance._toggleState(mapEvent);
        expect(instance.removeState).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('componentDidUpdate', () => {
    beforeEach(() => {
      mapDiv = document.createElement('div');
      mapDiv.setAttribute('id', 'tile-chart-map');
      window.domNode = mapDiv;
      document.body.appendChild(mapDiv);
    });

    afterEach(() => {
      const div = document.getElementById('tile-chart-map');
      if (div) {
        document.body.removeChild(div);
      }
      jest.clearAllMocks();
    });

    it('does nothing when no data', () => {
      target = shallow(<TileChartMap data={[]} />);
      redrawSpy = jest.spyOn(target.instance(), '_redrawMap');
      target.setProps({ data: [[]] });
      expect(TileMap).toHaveBeenCalledTimes(0);
    });

    it('redraw when the data is the same but map element is missing', () => {
      // append children to mock test
      target = shallow(<TileChartMap data={[[{ name: 'TX', value: 100 }]]} />);
      redrawSpy = jest.spyOn(target.instance(), '_redrawMap');
      target.setProps({ data: [[{ name: 'TX', value: 100 }]] });
      expect(TileMap).toHaveBeenCalledTimes(1);
      expect(redrawSpy).toHaveBeenCalledTimes(1);
    });

    it('skips redraw when the data is the same', () => {
      mapDiv.appendChild(document.createElement('foobar'));
      target = shallow(<TileChartMap data={[[23, 4, 3]]} />);
      redrawSpy = jest.spyOn(target.instance(), '_redrawMap');
      target.setProps({ data: [[23, 4, 3]] });
      expect(redrawSpy).toHaveBeenCalledTimes(0);
      expect(TileMap).toHaveBeenCalledTimes(0);
    });

    it('trigger a new update when data changes', () => {
      target = shallow(
        <TileChartMap
          data={[
            [
              { name: 'TX', value: 100 },
              { name: 'LA', value: 10 },
            ],
          ]}
          dataNormalization={'None'}
        />
      );
      redrawSpy = jest.spyOn(target.instance(), '_redrawMap');
      target.setProps({
        data: [
          [
            { name: 'TX', value: 100 },
            { name: 'LA', value: 100 },
          ],
        ],
      });
      expect(redrawSpy).toHaveBeenCalledTimes(1);
      expect(TileMap).toHaveBeenCalledTimes(1);
    });

    it('trigger a new update when printMode changes', () => {
      target = shallow(
        <TileChartMap
          data={[
            [
              { name: 'TX', value: 100 },
              { name: 'LA', value: 10 },
            ],
          ]}
          printMode={false}
        />
      );
      redrawSpy = jest.spyOn(target.instance(), '_redrawMap');
      target.setProps({ printMode: true });
      expect(redrawSpy).toHaveBeenCalledTimes(1);
      expect(TileMap).toHaveBeenCalledTimes(1);
    });

    it('trigger a new update when width changes', () => {
      target = shallow(
        <TileChartMap
          data={[
            [
              { name: 'TX', value: 100 },
              { name: 'LA', value: 10 },
            ],
          ]}
          printMode={false}
          width={1000}
        />
      );
      redrawSpy = jest.spyOn(target.instance(), '_redrawMap');
      target.setProps({ width: 600 });
      expect(redrawSpy).toHaveBeenCalledTimes(1);
      expect(TileMap).toHaveBeenCalledTimes(1);
    });
  });

  describe('mapDispatchToProps', () => {
    let dispatch, gaSpy;
    beforeEach(() => {
      jest.clearAllMocks();
      dispatch = jest.fn();
      gaSpy = jest.spyOn(utils, 'sendAnalyticsEvent');
    });
    it('provides a way to call addState', () => {
      mapDispatchToProps(dispatch).addState({ abbr: 'foo', name: 'bar' });
      expect(dispatch.mock.calls.length).toEqual(1);
      expect(gaSpy).toHaveBeenCalledWith('State Event: add', 'foo');
    });

    it('provides a way to call removeState', () => {
      mapDispatchToProps(dispatch).removeState({ abbr: 'foo', name: 'bar' });
      expect(dispatch.mock.calls.length).toEqual(1);
      expect(gaSpy).toHaveBeenCalledWith('State Event: remove', 'foo');
    });
  });

  describe('mapStateToProps', () => {
    it('maps state and props', () => {
      const state = {
        map: {
          results: {
            state: [
              // name comes from agg api
              {
                name: 'TX',
                issue: 'something',
                product: 'a prod',
                value: 100000,
              },
              { name: 'LA', issue: 'something', product: 'b prod', value: 2 },
              { name: 'CA', issue: 'something', product: 'c prod', value: 3 },
              { name: 'MH', issue: 'real data', product: 'is messy', value: 9 },
            ],
          },
        },
        query: {
          dataNormalization: false,
          state: ['TX'],
        },
        view: {
          printMode: false,
          width: 1000,
        },
      };
      let actual = mapStateToProps(state);
      expect(actual).toEqual({
        data: [
          [
            {
              abbr: 'TX',
              name: 'TX',
              fullName: 'Texas',
              className: 'selected',
              issue: 'something',
              perCapita: '3.65',
              product: 'a prod',
              value: 100000,
            },
            {
              abbr: 'LA',
              name: 'LA',
              fullName: 'Louisiana',
              className: 'deselected',
              issue: 'something',
              perCapita: '0.00',
              product: 'b prod',
              value: 2,
            },
            {
              abbr: 'CA',
              name: 'CA',
              fullName: 'California',
              className: 'deselected',
              issue: 'something',
              perCapita: '0.00',
              product: 'c prod',
              value: 3,
            },
            {
              abbr: 'MH',
              name: 'MH',
              fullName: '',
              className: 'deselected',
              issue: 'real data',
              perCapita: '9000.00',
              product: 'is messy',
              value: 9,
            },
          ],
        ],
        dataNormalization: false,
        hasTip: true,
        printClass: '',
        stateFilters: ['TX'],
        width: 1000,
      });
    });

    it('maps state and props - no filters', () => {
      const state = {
        map: {
          results: {
            state: [
              // name comes from agg api
              {
                name: 'TX',
                issue: 'something',
                product: 'a prod',
                value: 100000,
              },
              { name: 'LA', issue: 'something', product: 'b prod', value: 2 },
              { name: 'CA', issue: 'something', product: 'c prod', value: 3 },
              { name: 'MH', issue: 'real data', product: 'is messy', value: 9 },
            ],
          },
        },
        query: {
          dataNormalization: types.GEO_NORM_NONE,
        },
        view: {
          printMode: false,
          width: 1000,
        },
      };
      let actual = mapStateToProps(state);
      expect(actual).toEqual({
        data: [
          [
            {
              abbr: 'TX',
              name: 'TX',
              fullName: 'Texas',
              className: '',
              issue: 'something',
              perCapita: '3.65',
              product: 'a prod',
              value: 100000,
            },
            {
              abbr: 'LA',
              name: 'LA',
              fullName: 'Louisiana',
              className: '',
              issue: 'something',
              perCapita: '0.00',
              product: 'b prod',
              value: 2,
            },
            {
              abbr: 'CA',
              name: 'CA',
              fullName: 'California',
              className: '',
              issue: 'something',
              perCapita: '0.00',
              product: 'c prod',
              value: 3,
            },
            {
              abbr: 'MH',
              name: 'MH',
              fullName: '',
              className: '',
              issue: 'real data',
              perCapita: '9000.00',
              product: 'is messy',
              value: 9,
            },
          ],
        ],
        dataNormalization: types.GEO_NORM_NONE,
        hasTip: true,
        printClass: '',
        stateFilters: [],
        width: 1000,
      });
    });
  });
});
