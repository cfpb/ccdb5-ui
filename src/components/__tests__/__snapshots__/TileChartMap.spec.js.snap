// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: TileChartMap initial state renders print mode without crashing 1`] = `
<div>
  <div
    className="cfpb-chart print"
    data-chart-type="tile_map"
    id="tile-chart-map"
  />
</div>
`;

exports[`component: TileChartMap initial state renders without crashing 1`] = `
<div>
  <div
    className="cfpb-chart "
    data-chart-type="tile_map"
    id="tile-chart-map"
  />
</div>
`;
